SU-SWS/decanter
# D8CORE-2950: added link-icon-before

# READY FOR REVIEW

# Summary
We want mailto icons to display _before_ the icon. This adds link-icon-before() mixin to complement the existing link-icon() mixin. This new mixin works just like the original except it adds the icon to the front and reverses the animation.

# Needed By (Date)
Needed for https://stanfordits.atlassian.net/browse/D8CORE-2950

# Urgency
not urgent

# Steps to Test

1. Check out this branch
2. Create some mailto links
1. Verify it works as expected.

Here's the twig and HTML I used for testing

```
<div class="su-stuff">
  <h2>
    Caryl's Sample links
  </h2>
  <p>
    Various versions of links.
  </p>

  {% set link_vars = {
    modifier_class: 'focus-change',
    link_content: 'Example link',
    link_href: 'https://stanford.edu'
  } %}

  <p>
    This is a link{% include '@decanter/components/link/link.twig' with link_vars only %}inside
    of a sentence.
  </p>

  {% set link_vars = {
    modifier_class: 'focus-change',
    link_content: 'Example H2 link',
    link_href: 'https://stanford.edu'
  } %}


  {% set link_vars = {
    modifier_class: 'focus-change',
    link_content: 'Example mailto link',
    link_href: 'mailto:sws-developers@lists.stanford.edu'
  } %}

  <p>
    This is a link{% include '@decanter/components/link/link.twig' with link_vars only %}inside
    of a sentence.
  </p>

  {% set link_vars = {
    modifier_class: 'focus-change',
    link_content: 'Example H2 mailto link',
    link_href: 'mailto:sws-developers@lists.stanford.edu'
  } %}

  <h2>
    {% include '@decanter/components/link/link.twig' with link_vars only %}
  </h2>

  <h2>
    This is a link{% include '@decanter/components/link/link.twig' with link_vars only %}inside
    of a header
  </h2>

  {% set link_vars = {
    modifier_class: 'focus-change',
    link_content: 'Example H3 link',
    link_href: 'https://stanford.edu'
  } %}

  <h3>
    {% include '@decanter/components/link/link.twig' with link_vars only %}
  </h3>

  <h3>
    This is a link{% include '@decanter/components/link/link.twig' with link_vars only %}inside
    of a h3 header
  </h3>


  <h2>
    This is a default button
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button focus-change',
    link_content: 'Default Button link',
    link_href: 'https://stanford.edu'
  } %}

  {% include '@decanter/components/link/link.twig' with link_vars only %}

  <h2>
    This is a Big button
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button--big focus-change',
    link_content: 'Big Button link',
    link_href: 'https://stanford.edu'
  } %}
  {% include '@decanter/components/link/link.twig' with link_vars only %}

  <h2>
    This is a Secondary button
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button--secondary focus-change',
    link_content: 'Secondary Button link',
    link_href: 'https://stanford.edu'
  } %}
  {% include '@decanter/components/link/link.twig' with link_vars only %}

    <h3>
    This is a link{% include '@decanter/components/link/link.twig' with link_vars only %}inside
    of a h3 header
  </h3>


  <h2>
    This is a default button with a mailto link
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button focus-change',
    link_content: 'Default Button email',
    link_href: 'mailto:sws-developers@lists.stanford.edu'
  } %}

  {% include '@decanter/components/link/link.twig' with link_vars only %}

  <h2>
    This is a Big button with a mailto link
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button--big focus-change',
    link_content: 'Big Button email',
    link_href: 'mailto:sws-developers@lists.stanford.edu'
  } %}
  {% include '@decanter/components/link/link.twig' with link_vars only %}

  <h2>
    This is a Secondary button with a mailto link
  </h2>

  {% set link_vars = {
    modifier_class: 'su-button--secondary focus-change',
    link_content: 'Secondary Button email',
    link_href: 'mailto:sws-developers@lists.stanford.edu'
  } %}
  {% include '@decanter/components/link/link.twig' with link_vars only %}


 <h2>1 Hello World!</h2>
  <p>There are words in this paragraph</p>
  <h2 tabindex="0">Hello World!</h2>
  <p>There are words in this paragraph</p>
<a href="#">
  <h2 tabindex="0">Hello World!</h2>
  <p>There are words in this paragraph</p>

</div>

````
Here's the SASS as well
```
// Include the mailto icon.
a[href^='mailto:'] {
  // Text.
  @include link-icon(email-24px, 1em, null, null, before);

  &::before {
    font-size: 1em;
    margin-bottom: -0.2em;
    background-color: $su-color-bright-blue;
  }

  &:hover,
  &:focus,
  &:active {
    &::before {
      background-color: $su-color-black;
    }
  }
}
```

# Affected Projects or Products
- D8Core and other 

# Associated Issues and/or People
- JIRA ticket https://stanfordits.atlassian.net/browse/D8CORE-2950
- Other PRs: https://github.com/SU-SWS/stanford_profile_helper/pull/53
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
